File: backend/src/Tmap.Api/Features/Recurrence/RecurrenceDtos.cs
using Tmap.Api.Infrastructure.Entities;

namespace Tmap.Api.Features.Recurrence;

// ---- create(task + rule) ----
public sealed record RecurringTaskInput(
    string Title,
    string Notes,
    Guid? ProjectId,
    List<string> Labels,
    string Source,
    DateOnly? PlannedDate,
    int DurationMinutes,
    int? Priority,
    int? ReminderMinutes,
    Guid? Id = null);

public sealed record RecurrenceRuleInput(
    RecurrenceFrequency Frequency,
    int Interval,
    List<int> DaysOfWeek,
    RecurrenceEndType EndType,
    int? EndCount,
    DateOnly? EndDate);

public sealed record CreateRecurringTaskRequest(
    RecurringTaskInput Task,
    RecurrenceRuleInput Rule);


// ---- rule read ----
public sealed record RecurrenceRuleResponse(
    Guid Id,
    RecurrenceFrequency Frequency,
    int Interval,
    List<int> DaysOfWeek,
    RecurrenceEndType EndType,
    int? EndCount,
    DateOnly? EndDate,
    DateOnly? GeneratedUntil,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt);

// ---- updateRule (whole-rule replace of mutable fields) ----
public sealed record UpdateRuleRequest(
    RecurrenceFrequency Frequency,
    int Interval,
    List<int> DaysOfWeek,
    RecurrenceEndType EndType,
    int? EndCount,
    DateOnly? EndDate);

// ---- updateSeries(ruleId, updates): field-level edits propagated to template + future live instances ----
public sealed record UpdateSeriesRequest(
    string? Title,
    string? Notes,
    Guid? ProjectId,
    List<string>? Labels,
    int? DurationMinutes,
    int? Priority,
    int? ReminderMinutes);

// ---- deleteSeriesFuture(ruleId, fromDate) ----
public sealed record DeleteSeriesFutureRequest(DateOnly FromDate);

File: android/app/src/main/java/net/qmindtech/tmap/data/remote/dto/RecurrenceDtos.kt
package net.qmindtech.tmap.data.remote.dto

import kotlinx.serialization.Serializable

/**
 * Request/response DTOs for the /api/v1/recurrence endpoints. Field names + types mirror the backend
 * records verbatim (RecurrenceDtos.cs). Enum-valued fields (frequency/endType) are plain Strings
 * carrying the PascalCase wire values ("Daily"/"Weekly", "Never"/"Count"/"Date").
 */

@Serializable
data class RecurringTaskInput(
    val title: String,
    val notes: String,
    val projectId: String?,
    val labels: List<String>,
    val source: String,
    val plannedDate: String?, // yyyy-MM-dd; null => server anchors to today
    val durationMinutes: Int,
    val priority: Int?,
    val reminderMinutes: Int?,
    val id: String? = null,
)

@Serializable
data class RecurrenceRuleInput(
    val frequency: String, // "Daily" | "Weekly"
    val interval: Int,
    val daysOfWeek: List<Int>, // Sunday=0..6; [] for daily
    val endType: String, // "Never" | "Count" | "Date"
    val endCount: Int?,
    val endDate: String?, // yyyy-MM-dd
    val id: String? = null,
)

@Serializable
data class CreateRecurringTaskRequest(
    val task: RecurringTaskInput,
    val rule: RecurrenceRuleInput,
)

/** PATCH /recurrence/rules/{id} — whole-rule replace (id comes from the route). */
@Serializable
data class UpdateRuleRequest(
    val frequency: String,
    val interval: Int,
    val daysOfWeek: List<Int>,
    val endType: String,
    val endCount: Int?,
    val endDate: String?,
)

/** POST /recurrence/rules/{id}/delete-future body. */
@Serializable
data class DeleteSeriesFutureRequest(
    val fromDate: String, // yyyy-MM-dd, required
)

/**
 * INTERNAL outbox envelope for a RECURRENCE + OpType.DELETE op (not a wire body). PushRunner reads
 * `scope` to choose DELETE /rules/{id} (scope="all") vs POST /rules/{id}/delete-future (scope="future").
 */
@Serializable
data class RecurrenceDeletePayload(
    val scope: String, // "all" | "future"
    val fromDate: String? = null, // required when scope == "future"
)
